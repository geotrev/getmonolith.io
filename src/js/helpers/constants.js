export const KeyCodes = {
  SHIFT: 16,
  TAB: 9,
  ARROW_UP: 38,
  ARROW_DOWN: 40,
}

export const Selectors = {
  FOCUSABLE_TAGS: ["a", "button", "input", "object", "select", "textarea", "[tabindex]"],
  KEYBOARD_CLASS: "using-keyboard",
  NOT_VISUALLY_HIDDEN_CLASS: ":not(.is-visually-hidden)",
  TABINDEX: "tabindex",
}

export const Events = {
  KEYDOWN: "keydown",
  CLICK: "click",
  BLUR: "blur",
}

export const Messages = {
  NO_SELECTOR_STRING_OR_CHILDREN_ERROR:
    "createFocusTrap must be given one or both of: first parameter (as selector string)" +
    " and/or options.children (array of elements).",
  OPTION_MATCHERS_DATA_TYPE_ERROR:
    "Invalid data type given to options.matchers for createFocusTrap. Expected: Array.",
  INCORRECT_MATCHER_TYPE_ERROR: type =>
    `Invalid matcher given to options.matchers for createFocusTrap. Expected: String. Recieved: ${type}.`,
  NO_MATCHER_LENGTH_ERROR:
    "Invalid value given to options.matchers for createFocusTrap; value must be an array with at least one selector string",
}
