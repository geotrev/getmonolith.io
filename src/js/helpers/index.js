export * from "./component-engine"
export { createFocusRing } from "./create-focus-ring"
export { createFocusTrap } from "./create-focus-trap"
export * from "./dom"
export { focusOnce } from "./focus-once"
export { isBrowserEnv } from "./is-browser-env"
export { isiOSMobile } from "./is-ios-mobile"
export { log } from "./log"
export { throttle } from "./throttle"
