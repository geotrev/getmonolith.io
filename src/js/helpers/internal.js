/**
 * Private method and property symbols used in the base class, mixins, and components.
 * These are used to prevent external scripts from accessing the internals of Undernet's
 * core logic.
 */
export const internal = {
  // Component
  start: Symbol("start"),
  stop: Symbol("stop"),
  validate: {
    method: Symbol("validate"),
    property: Symbol("_validate"),
  },
  teardown: {
    method: Symbol("teardown"),
    property: Symbol("_teardown"),
  },
  components: {
    method: Symbol("components"),
    property: Symbol("_components"),
  },
  activeComponent: {
    method: Symbol("activeComponent"),
    property: Symbol("_activeComponent"),
  },
  cancelActiveComponent: {
    method: Symbol("cancelActiveComponent"),
    property: Symbol("_cancelActiveComponent"),
  },
  // Mixins
  isVisible: {
    method: Symbol("isVisible"),
    property: Symbol("_isVisible"),
  },
}
