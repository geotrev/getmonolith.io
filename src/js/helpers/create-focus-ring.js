import { Events, Selectors } from "./constants"
import { isBrowserEnv } from "./is-browser-env"

/**
 * Factory function that creates focus ring helpers.
 *
 * ```js
 * const focusRing = createFocusRing()
 * focusRing.start()
 * focusRing.stop()
 * ```
 *
 * @returns {{ start: Function, stop: Function }}
 */
export const createFocusRing = () => {
  if (!isBrowserEnv) return

  let listeningForKeydown

  const listenForKeyboard = () => {
    document.body.classList.add(Selectors.KEYBOARD_CLASS)
    document.removeEventListener(Events.KEYDOWN, listenForKeyboard)
    // eslint-disable-next-line no-use-before-define
    document.addEventListener(Events.CLICK, listenForClick)
    listeningForKeydown = false
  }

  const listenForClick = () => {
    document.body.classList.remove(Selectors.KEYBOARD_CLASS)
    document.removeEventListener(Events.CLICK, listenForClick)
    document.addEventListener(Events.KEYDOWN, listenForKeyboard)
    listeningForKeydown = true
  }

  return {
    start() {
      document.addEventListener(Events.KEYDOWN, listenForKeyboard)
    },
    stop() {
      if (listeningForKeydown) {
        document.removeEventListener(Events.KEYDOWN, listenForKeyboard)
      } else {
        document.body.classList.remove(Selectors.KEYBOARD_CLASS)
        document.removeEventListener(Events.CLICK, listenForClick)
      }
    },
  }
}
