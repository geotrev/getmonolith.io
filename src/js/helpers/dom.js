import { Selectors } from "./constants"

/**
 * Run a `querySelectorAll` but convert the result to an array literal for IE11 compat.
 * @param {string} selector
 * @param {HTMLElement} parent
 */
export const queryAll = (selector, parent = document) => {
  return Array.apply(null, parent.querySelectorAll(selector))
}

/**
 * Search for elements matching a given selector.
 *
 * ```js
 * const elements = getFocusableElements(".wrapper")
 * elements.forEach(element => element.classList.add("focusable"))
 * ```
 *
 * Use a custom pattern to select only specific elements:
 *
 * ```js
 * const elements = getFocusableElements(".wrapper", [".my-button"])
 * ```
 *
 * @param {String} selectorString - The selector string of the container element.
 * @param {String[]} matchers - Optional matchers override. Defaults to common focusable selectors.
 * @returns {Element[]} Static array of HTML elements
 */
export const getFocusableElements = (selectorString, matchers = Selectors.FOCUSABLE_TAGS) => {
  const focusables = matchers
    .map(selector => `${selectorString} ${selector}${Selectors.NOT_VISUALLY_HIDDEN_CLASS}`)
    .join(", ")

  return queryAll(focusables)
}
