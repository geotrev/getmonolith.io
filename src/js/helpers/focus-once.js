import { Events, Selectors } from "./constants"

/**
 * Focus a single element one time and teardown when unfocused.
 * @param {HTMLElement} element
 */
export const focusOnce = element => {
  const handleBlur = ({ target }) => {
    target.removeAttribute(Selectors.TABINDEX)
    target.removeEventListener(Events.BLUR, handleBlur)
  }

  element.setAttribute(Selectors.TABINDEX, "-1")
  element.focus()
  element.addEventListener(Events.BLUR, handleBlur)
}
