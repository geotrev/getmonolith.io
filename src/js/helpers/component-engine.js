import { queryAll } from "./dom"
import { isBrowserEnv } from "./is-browser-env"

const isString = value => typeof value === "string"
const isFunction = value => typeof value === "function"

export const ComponentEngine = {
  /**
   * Initialize an Undernet component globally or by id.
   * @param {Object} metadata
   * @property {string} id
   * @property {string} attribute
   * @property {Object} thisArg
   */
  start: (metadata = {}) => {
    if (!isBrowserEnv) return

    const { id, attribute, thisArg } = metadata

    if (id && isString(id)) {
      const instance = document.querySelector(`[${attribute}='${id}']`)
      if (!instance) return

      const validComponent = [instance].filter(thisArg._validate)[0]
      if (!validComponent) return

      thisArg._components.push(validComponent)
    } else if (!id && !thisArg._components.length) {
      const instances = queryAll(`[${attribute}]`)
      if (!instances.length) return

      const validComponents = instances.filter(thisArg._validate)
      thisArg._components = thisArg._components.concat(validComponents)
    } else {
      // attempted to .start() when .stop() wasn't run,
      // OR tried to instantiate a component that's already active.
    }
  },
  /**
   * Teardown an Undernet component globally or by id.
   * @param {Object} metadata
   * @property {string} id
   * @property {string} attribute
   * @property {Object} thisArg
   * @property {string} activeNodeKey
   * @property {*=} cancelActiveFn
   */
  stop: (metadata = {}) => {
    if (!isBrowserEnv) return

    const { id, attribute, thisArg, activeNodeKey, cancelActiveFn } = metadata

    if (id && isString(id)) {
      let targetIndex
      const instance = thisArg._components.filter((activeInstance, index) => {
        if (activeInstance.getAttribute(attribute) !== id) return false
        targetIndex = index
        return true
      })[0]

      if (!instance) return

      if (
        thisArg[activeNodeKey] &&
        instance === thisArg[activeNodeKey] &&
        isFunction(cancelActiveFn)
      )
        thisArg[cancelActiveFn]()

      thisArg._teardown(instance)
      thisArg._components.splice(targetIndex, 1)
    } else if (!id && thisArg._components.length) {
      if (thisArg[activeNodeKey] && isFunction(cancelActiveFn)) thisArg[cancelActiveFn]()

      thisArg._components.forEach(thisArg._teardown)
      thisArg._components = []
    }
  },
}
