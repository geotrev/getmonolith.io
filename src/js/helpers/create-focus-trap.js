import { isBrowserEnv } from "./is-browser-env"
import { Selectors, Messages, KeyCodes, Events } from "./constants"
import { log } from "./log"
import { getFocusableElements } from "./dom"
/**
 * Factory function that creates focus trap helpers.
 *
 * ```js
 * const focusTrap = createFocusTrap("#element-id")
 * focusTrap.start()
 * focusTrap.stop()
 * ```
 *
 * Pass an object in the second param to use other helpers for arrow trapping, custom children, and custom matchers.
 *
 * ```js
 * const focusTrap = createFocusTrap("#element-id", {
 *    useArrows: true,
 *    children: [],
 *    matchers: ['a', 'button', '.my-cool-element']
 * })
 * focusTrap.start()
 * ```
 *
 * @param {String} selectorString
 * @param {{ useArrows, children: (NodeList|Element[]), matchers: String[] }} options - useArrows is coerced to true/false.
 * @returns {{ start: Function, stop: Function }}
 */
export const createFocusTrap = (selectorString, options = {}) => {
  if (!isBrowserEnv) return
  let focusTrapActive = false
  const { useArrows, children, matchers = Selectors.FOCUSABLE_TAGS } = options

  if (!selectorString && !children.length) {
    log(Messages.NO_SELECTOR_STRING_OR_CHILDREN_ERROR)
    return
  }

  if (!Array.isArray(matchers)) {
    log(Messages.OPTION_MATCHERS_DATA_TYPE_ERROR)
    return
  } else if (matchers.length) {
    let hasBadMatcher = false

    matchers.forEach(matcher => {
      const type = typeof matcher
      if (type !== "string") {
        log(Messages.INCORRECT_MATCHER_TYPE_ERROR(type))
        hasBadMatcher = true
      }
    })

    if (hasBadMatcher) return
  } else if (!matchers.length) {
    log(Messages.NO_MATCHER_LENGTH_ERROR)
    return
  }

  const focusableChildren =
    children && children.length ? children : getFocusableElements(selectorString, matchers)
  const focusableFirstChild = focusableChildren[0]
  const focusableLastChild = focusableChildren[focusableChildren.length - 1]

  const handleFocusTrapWithTab = event => {
    const containerElement = document.querySelector(selectorString)
    const containerActive = document.activeElement === containerElement
    const firstActive = document.activeElement === focusableFirstChild
    const lastActive = document.activeElement === focusableLastChild

    if (!containerActive && !firstActive && !lastActive) return

    const tabKey = event.which === KeyCodes.TAB
    const shiftKey = event.which === KeyCodes.SHIFT || event.shiftKey
    const hasShift = shiftKey && tabKey
    const noShift = !shiftKey && tabKey

    if (hasShift && (firstActive || containerActive)) {
      event.preventDefault()
      focusableLastChild.focus()
    } else if (noShift && lastActive) {
      event.preventDefault()
      focusableFirstChild.focus()
    }
  }

  const focusNextChild = () => {
    for (let i = 0; i < focusableChildren.length; i++) {
      if (focusableChildren[i] === document.activeElement) {
        focusableChildren[i + 1].focus()
        break
      }
    }
  }

  const focusLastChild = () => {
    for (let i = 0; i < focusableChildren.length; i++) {
      if (focusableChildren[i] === document.activeElement) {
        focusableChildren[i - 1].focus()
        break
      }
    }
  }

  const handleFocusTrapWithArrows = event => {
    const firstActive = document.activeElement === focusableFirstChild
    const lastActive = document.activeElement === focusableLastChild
    const arrowUp = event.which === KeyCodes.ARROW_UP
    const arrowDown = event.which === KeyCodes.ARROW_DOWN

    if (arrowUp || arrowDown) {
      event.preventDefault()

      if (firstActive && arrowUp) {
        focusableLastChild.focus()
      } else if (lastActive && arrowDown) {
        focusableFirstChild.focus()
      } else if (arrowDown) {
        focusNextChild()
      } else if (arrowUp) {
        focusLastChild()
      }
    }
  }

  return {
    start() {
      // Already active, exit
      if (focusTrapActive) return

      // Now active
      focusTrapActive = true

      // Handle focus listeners
      if (useArrows) {
        document.addEventListener(Events.KEYDOWN, handleFocusTrapWithArrows)
      } else {
        document.addEventListener(Events.KEYDOWN, handleFocusTrapWithTab)
      }
    },
    stop() {
      // Nothing to stop, exit
      if (!focusTrapActive) return

      // Now inactive
      focusTrapActive = false

      // Remove focus listeners
      if (useArrows) {
        document.removeEventListener(Events.KEYDOWN, handleFocusTrapWithArrows)
      } else {
        document.removeEventListener(Events.KEYDOWN, handleFocusTrapWithTab)
      }
    },
  }
}
