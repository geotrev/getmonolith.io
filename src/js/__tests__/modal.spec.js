import Undernet from "../"
import { find, renderDOM, simulateKeyboardEvent } from "./helpers"

const dom = `
  <button data-target="modal-id">Open modal</button>

  <div className="modal-overlay" data-modal="modal-id">
    <div className="modal-dialog" data-parent="modal-id" aria-labelledby="header-id">
      <header>
        <h2 className="h6 has-no-m-block-start" id="header-id">
          Modal Header
        </h2>
        <a data-close href="#">
          <span aria-hidden="true">&times;</span>
        </a>
      </header>
      <section>
        <p>Some modal content here</p>
      </section>
      <footer>
        <a className="button" data-close href="#">
          Cancel
        </a>
        <a className="primary button" href="#">
          OK
        </a>
      </footer>
    </div>
  </div>
`

const KeyCodes = {
  ESCAPE: 27,
}

const activeElement = () => document.activeElement

describe("Modal", () => {
  afterEach(() => {
    Undernet.Modals.stop()
  })

  describe("API start", () => {
    it("sets attributes", () => {
      // Given
      const wrapper = renderDOM(dom)
      // When
      Undernet.Modals.start()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("API stop + #handleClick", () => {
    it("does not open modal", () => {
      // Given
      const wrapper = renderDOM(dom)
      const trigger = find("[data-target='modal-id']")
      // When
      Undernet.Modals.start()
      Undernet.Modals.stop()
      trigger.click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("#handleClick", () => {
    let wrapper

    beforeEach(() => {
      wrapper = renderDOM(dom)

      const trigger = find("[data-target='modal-id']")

      Undernet.Modals.start()
      trigger.click()
    })

    it("opens modal", () => {
      expect(wrapper()).toMatchSnapshot()
    })

    it("sets focus to modal dialog", () => {
      const modalDialog = find("[data-parent]")
      expect(activeElement()).toEqual(modalDialog)
    })
  })

  describe("#handleClose", () => {
    let wrapper
    let trigger
    let closeButton

    beforeEach(() => {
      wrapper = renderDOM(dom)

      trigger = find("[data-target='modal-id']")
      closeButton = find("header [data-close]")

      Undernet.Modals.start()
      trigger.click()
      closeButton.click()
    })

    it("closes modal", () => {
      expect(wrapper()).toMatchSnapshot()
    })

    it("sets focus to modal trigger", () => {
      expect(activeElement()).toEqual(trigger)
    })

    it("removes tabindex when trigger loses focus", () => {
      trigger.blur()
      expect(trigger.hasAttribute("tabindex")).toEqual(false)
    })
  })

  describe("#handleOverlayClick", () => {
    it("closes modal", () => {
      // Given
      const wrapper = renderDOM(dom)
      const trigger = find("[data-target='modal-id']")
      // When
      Undernet.Modals.start()
      trigger.click()
      find("[data-modal='modal-id']").click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("#handleEscapeKeyPress", () => {
    it("closes modal", () => {
      // Given
      const wrapper = renderDOM(dom)
      const trigger = find("[data-target='modal-id']")
      // When
      Undernet.Modals.start()
      trigger.click()
      simulateKeyboardEvent(KeyCodes.ESCAPE)
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("#handleReturnFocus", () => {
    it("sets focus back to modal trigger", () => {
      // Given
      renderDOM(dom)
      const trigger = find("[data-target='modal-id']")
      const closeButton = find("header [data-close]")
      // When
      Undernet.Modals.start()
      trigger.click()
      closeButton.click()
      // Then
      expect(activeElement()).toEqual(trigger)
    })
  })

  describe("#handleScrollRestore", () => {
    let trigger
    let closeButton

    beforeEach(() => {
      renderDOM(dom)

      trigger = find("[data-target='modal-id']")
      closeButton = find("header [data-close]")

      Undernet.Modals.start()
      trigger.click()
      closeButton.click()
    })

    it("removes 'no-scroll' class from <body>", () => {
      expect(find("body").className).toEqual("")
    })

    it("removes 'no-scroll' class from <html>", () => {
      expect(find("html").className).toEqual("")
    })
  })

  describe("#handleScrollStop", () => {
    beforeEach(() => {
      renderDOM(dom)
      const trigger = find("[data-target='modal-id']")

      Undernet.Modals.start()
      trigger.click()
    })

    it("sets 'no-scroll' class to <body>", () => {
      expect(find("body").className).toEqual("no-scroll")
    })

    it("sets 'no-scroll' class to <html>", () => {
      expect(find("html").className).toEqual("no-scroll")
    })
  })
})

const errorDom = (target, modal, parent) => `
  <button data-target="${target}">Open modal</button>

  <div className="modal-overlay" data-modal="${modal}">
    <div className="modal-dialog" data-parent="${parent}" aria-labelledby="header-id">
      <header>
        <h2 className="h6 has-no-m-block-start" id="header-id">
          Modal Header
        </h2>
        <a data-close href="#">
          <span aria-hidden="true">&times;</span>
        </a>
      </header>
      <section>
        <p>Some modal content here</p>
      </section>
      <footer>
        <a className="button" data-close href="#">
          Cancel
        </a>
        <a className="primary button" href="#">
          OK
        </a>
      </footer>
    </div>
  </div>
`

describe("Modal Warnings", () => {
  beforeAll(() => {
    console.warning = jest.fn()
  })

  it("throws error if modal button isn't found", () => {
    // Given
    renderDOM(errorDom("", "modal-id", "modal-id"))
    // When
    Undernet.Modals.start()
    // Then
    expect(console.warning).toBeCalledWith("Could not find modal trigger with id modal-id.")
  })

  it("throws error if modal id isn't found", () => {
    // Given
    renderDOM(errorDom("modal-id", "", "modal-id"))
    // When
    Undernet.Modals.start()
    // Then
    expect(console.warning).toBeCalledWith(
      "Could not detect an id on your [data-modal] element. Please add a value matching the modal trigger's [data-parent] attribute."
    )
  })

  it("throws error if [data-parent] attribute does not match its parent [data-modal]", () => {
    // Given
    renderDOM(errorDom("modal-id", "modal-id", ""))
    // When
    Undernet.Modals.start()
    // Then
    expect(console.warning).toBeCalledWith(
      `Could not find element with attribute [data-parent='modal-id'].`
    )
  })
})
