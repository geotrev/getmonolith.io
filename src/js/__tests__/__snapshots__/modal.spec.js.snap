// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modals #handleClose -> Modal Close Button Click closes modal 1`] = `
<body
  class=""
  style="padding-right: 1024px;"
>
  
  
  <button
    data-target="new-modal"
  >
    Open modal
  </button>
  

  
  <div
    aria-hidden="true"
    classname="modal-overlay"
    data-modal="new-modal"
    data-visible="false"
    style="padding-left: 1024px;"
  >
    
    
    <div
      aria-labelledby="header-id"
      aria-modal="true"
      classname="modal-dialog"
      data-parent="new-modal"
      role="dialog"
    >
      
      
      <header>
        
        
        <h2
          classname="h6 has-no-margin-top"
          id="header-id"
        >
          
          Modal Header
        
        </h2>
        
        
        <a
          data-close=""
          href="#"
          tabindex="-1"
        >
          
          
          <span
            aria-hidden="true"
          >
            ×
          </span>
          
        
        </a>
        
      
      </header>
      
      
      <section>
        
        
        <p>
          Some modal content here
        </p>
        
      
      </section>
      
      
      <footer>
        
        
        <a
          classname="button"
          data-close=""
          href="#"
          tabindex="-1"
        >
          
          Cancel
        
        </a>
        
        
        <a
          classname="primary button"
          href="#"
          tabindex="-1"
        >
          
          OK
        
        </a>
        
      
      </footer>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Modals #handleEscapeKeyPress -> Escape Key Press closes modal 1`] = `
<div
  aria-hidden="true"
  classname="modal-overlay"
  data-modal="new-modal"
  data-visible="false"
  style="padding-left: 1024px;"
>
  
    
  <div
    aria-labelledby="header-id"
    aria-modal="true"
    classname="modal-dialog"
    data-parent="new-modal"
    role="dialog"
  >
    
      
    <header>
      
        
      <h2
        classname="h6 has-no-margin-top"
        id="header-id"
      >
        
          Modal Header
        
      </h2>
      
        
      <a
        data-close=""
        href="#"
        tabindex="-1"
      >
        
          
        <span
          aria-hidden="true"
        >
          ×
        </span>
        
        
      </a>
      
      
    </header>
    
      
    <section>
      
        
      <p>
        Some modal content here
      </p>
      
      
    </section>
    
      
    <footer>
      
        
      <a
        classname="button"
        data-close=""
        href="#"
        tabindex="-1"
      >
        
          Cancel
        
      </a>
      
        
      <a
        classname="primary button"
        href="#"
        tabindex="-1"
      >
        
          OK
        
      </a>
      
      
    </footer>
    
    
  </div>
  
  
</div>
`;

exports[`Modals #handleOverlayClick -> Modal Overlay Click closes modal 1`] = `
<div
  aria-hidden="true"
  classname="modal-overlay"
  data-modal="new-modal"
  data-visible="false"
  style="padding-left: 1024px;"
>
  
    
  <div
    aria-labelledby="header-id"
    aria-modal="true"
    classname="modal-dialog"
    data-parent="new-modal"
    role="dialog"
  >
    
      
    <header>
      
        
      <h2
        classname="h6 has-no-margin-top"
        id="header-id"
      >
        
          Modal Header
        
      </h2>
      
        
      <a
        data-close=""
        href="#"
        tabindex="-1"
      >
        
          
        <span
          aria-hidden="true"
        >
          ×
        </span>
        
        
      </a>
      
      
    </header>
    
      
    <section>
      
        
      <p>
        Some modal content here
      </p>
      
      
    </section>
    
      
    <footer>
      
        
      <a
        classname="button"
        data-close=""
        href="#"
        tabindex="-1"
      >
        
          Cancel
        
      </a>
      
        
      <a
        classname="primary button"
        href="#"
        tabindex="-1"
      >
        
          OK
        
      </a>
      
      
    </footer>
    
    
  </div>
  
  
</div>
`;

exports[`Modals #render -> Modal Button Click opens modal 1`] = `
<body
  class="no-scroll"
  style="padding-right: 1024px;"
>
  
  
  <button
    data-target="new-modal"
  >
    Open modal
  </button>
  

  
  <div
    aria-hidden="false"
    classname="modal-overlay"
    data-modal="new-modal"
    data-visible="true"
  >
    
    
    <div
      aria-labelledby="header-id"
      aria-modal="true"
      classname="modal-dialog"
      data-parent="new-modal"
      role="dialog"
      tabindex="-1"
    >
      
      
      <header>
        
        
        <h2
          classname="h6 has-no-margin-top"
          id="header-id"
        >
          
          Modal Header
        
        </h2>
        
        
        <a
          data-close=""
          href="#"
          tabindex="0"
        >
          
          
          <span
            aria-hidden="true"
          >
            ×
          </span>
          
        
        </a>
        
      
      </header>
      
      
      <section>
        
        
        <p>
          Some modal content here
        </p>
        
      
      </section>
      
      
      <footer>
        
        
        <a
          classname="button"
          data-close=""
          href="#"
          tabindex="0"
        >
          
          Cancel
        
        </a>
        
        
        <a
          classname="primary button"
          href="#"
          tabindex="0"
        >
          
          OK
        
        </a>
        
      
      </footer>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Modals API start sets attributes 1`] = `
<body>
  
  
  <button
    data-target="new-modal"
  >
    Open modal
  </button>
  

  
  <div
    aria-hidden="true"
    classname="modal-overlay"
    data-modal="new-modal"
    data-visible="false"
  >
    
    
    <div
      aria-labelledby="header-id"
      aria-modal="true"
      classname="modal-dialog"
      data-parent="new-modal"
      role="dialog"
    >
      
      
      <header>
        
        
        <h2
          classname="h6 has-no-margin-top"
          id="header-id"
        >
          
          Modal Header
        
        </h2>
        
        
        <a
          data-close=""
          href="#"
          tabindex="-1"
        >
          
          
          <span
            aria-hidden="true"
          >
            ×
          </span>
          
        
        </a>
        
      
      </header>
      
      
      <section>
        
        
        <p>
          Some modal content here
        </p>
        
      
      </section>
      
      
      <footer>
        
        
        <a
          classname="button"
          data-close=""
          href="#"
          tabindex="-1"
        >
          
          Cancel
        
        </a>
        
        
        <a
          classname="primary button"
          href="#"
          tabindex="-1"
        >
          
          OK
        
        </a>
        
      
      </footer>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Modals API stop -> Modal Button Click does not open modal 1`] = `
<body>
  
  
  <button
    data-target="new-modal"
  >
    Open modal
  </button>
  

  
  <div
    aria-hidden="true"
    classname="modal-overlay"
    data-modal="new-modal"
    data-visible="false"
  >
    
    
    <div
      aria-labelledby="header-id"
      aria-modal="true"
      classname="modal-dialog"
      data-parent="new-modal"
      role="dialog"
    >
      
      
      <header>
        
        
        <h2
          classname="h6 has-no-margin-top"
          id="header-id"
        >
          
          Modal Header
        
        </h2>
        
        
        <a
          data-close=""
          href="#"
          tabindex="-1"
        >
          
          
          <span
            aria-hidden="true"
          >
            ×
          </span>
          
        
        </a>
        
      
      </header>
      
      
      <section>
        
        
        <p>
          Some modal content here
        </p>
        
      
      </section>
      
      
      <footer>
        
        
        <a
          classname="button"
          data-close=""
          href="#"
          tabindex="-1"
        >
          
          Cancel
        
        </a>
        
        
        <a
          classname="primary button"
          href="#"
          tabindex="-1"
        >
          
          OK
        
        </a>
        
      
      </footer>
      
    
    </div>
    
  
  </div>
  

</body>
`;
